### PR TITLE
bt-690:  Make changes in search-users endpoints

### LINKED ISSUE #690: Make changes in search-users endpoints
-  Add search_type column to user-details table in DB
 -  add badges and hardskills to the response
 -  show only candidate, which were approved by admin
 -  change code in repository accorditng to task #605 
 -  change sorting by date from sorting by creation date to sorting by publish date

Image

### LINKED ISSUE #754: [BUG]The search using the filter “Years of experience” doesn’t correspond to the acceptance criteria
**Description:**
The search using the filter “Years of experience” doesn’t correspond to the acceptance criteria

**Preconditions:**
open http://bsa-talents.us-east-1.elasticbeanstalk.com/
login as Employer with valid data
open page http://bsa-talents.us-east-1.elasticbeanstalk.com/candidates

**Environment:**
OS - Windows
Browser - Google Chrome
Version - 116.0.5845.142

**Steps to reproduce:**

1. Choose the value “5+” in the filter “Years of experience”
2. Pay attention that as a result will return Talent’s cards, where Talents have less than 5+ years of experience
3. Choose several values in the filter “Years of experience”
4. Pay attention to the validation error notification that appears

**Actual Result:**

1. As a result of the search query returned Talent’s cards, that don’t correspond used filter
2. The user can't choose several values in the filter dropdown component

**Expected Result:**

1. As a result of the search query returned Talent’s cards, which corresponded to the used filter
2. The User can choose several values in the filter dropdown component

**Visual Proof (screenshots, videos, text):**
https://drive.google.com/file/d/1npfKjW5bVdX_nMrAcL2Rd26nhN-cluW5/view?usp=drive_link

**Severity:**
Blocker (S1)

### LINKED ISSUE #797: Remove tab "feedback" from preview page
This information is not viewable or editable for the Talent

Image

### PR DESCRIPTION
**Pull request by MChornomaz**
closes #690 
#754 #797
Add search_type column to user-details table in DB
add badges and hardskills to the response
show only candidate, which were approved by admin
change code in repository accorditng to task https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/issues/605
change sorting by date from sorting by creation date to sorting by publish date
image

### CODE REVIEW

File: backend/src/bundles/user-details/user-details.service.ts [hunk 1]
```diff
@@ -107,15 +110,35 @@ class UserDetailsService implements Service {
...
+
+            const userHardSkills = await getUserHardSkills(userDetailsId);
+
+            const userBadges = await getUserBadges(userDetailsId);
+
+            userDetails.hardSkills = userHardSkills;
+            userDetails.badges = userBadges;
+
+            return userDetails;
+        });
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/user-details.service.ts]: Why we can't fetch users hard skills and badges using withGraphFetched method for example and in this case we join data using sql raw query and it much better than async map with Promise.all

File: backend/src/bundles/user-details/helpers/apply-filters/helpers/apply-type-search-filter.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import { type UserDetailsModel } from '../../../user-details.model.js';
+
+const applyTypeSearchFilter = (
+    builder: QueryBuilder<UserDetailsModel, UserDetailsModel[]>,
+    payload: UserDetailsSearchUsersRequestDto,
+): void => {
+    if (payload.searchType && payload.searchType === SearchType.ACTIVE) {
+        void builder.where('searchType', payload.searchType);
+    }
+};
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/helpers/apply-filters/helpers/apply-type-search-filter.ts]: In this helper you use only payload.searchType, so you don't need to provide all payload to this helper

File: backend/src/bundles/user-details/helpers/apply-filters/helpers/apply-profile-name-filter.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import { type UserDetailsSearchUsersRequestDto } from '../../../types/types.js';
+import { type UserDetailsModel } from '../../../user-details.model.js';
+
+const applyProfileNameFilter = (
+    builder: QueryBuilder<UserDetailsModel, UserDetailsModel[]>,
+    payload: UserDetailsSearchUsersRequestDto,
+): void => {
+    if (payload.searchValue) {
+        void builder.where('profile_name', 'ilike', `%${payload.searchValue}%`);
+    }
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/helpers/apply-filters/helpers/apply-profile-name-filter.ts]: The same provide only data which you will use

File: backend/src/bundles/user-details/helpers/apply-filters/helpers/apply-location-filter.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+const applyLocationFilter = (
+    builder: QueryBuilder<UserDetailsModel, UserDetailsModel[]>,
+    payload: UserDetailsSearchUsersRequestDto,
+): void => {
+    if (payload.location) {
+        void builder.where((subquery) => {
+            searchByColumnValues(subquery, 'location', payload.location);
+        });
+    }
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/helpers/apply-filters/helpers/apply-location-filter.ts]: The same provide only data which you will use

File: backend/src/bundles/user-details/helpers/apply-filters/helpers/apply-job-title-filter.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+const applyJobTitleFilter = (
+    builder: QueryBuilder<UserDetailsModel, UserDetailsModel[]>,
+    payload: UserDetailsSearchUsersRequestDto,
+): void => {
+    if (payload.jobTitle && payload.jobTitle.length > 0) {
+        void builder.where((subquery) => {
+            searchByColumnValues(subquery, 'jobTitle', payload.jobTitle);
+        });
+    }
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/helpers/apply-filters/helpers/apply-job-title-filter.ts]: The same provide only data which you will use

File: backend/src/bundles/user-details/helpers/apply-filters/helpers/apply-hard-skills-filter.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+): void => {
+    if (payload.hardSkills && payload.hardSkills.length > 0) {
+        searchUserByRelativeTable({
+            builder,
+            values: payload.hardSkills,
+            columnName: 'hard_skill_id',
+            relativeTable: 'talentHardSkills',
+            alias: 'ths',
+        });
+    }
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/helpers/apply-filters/helpers/apply-hard-skills-filter.ts]: The same provide only data which you will use

File: backend/src/bundles/user-details/helpers/apply-filters/helpers/apply-experience-filter.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+const applyExperienceFilter = (
+    builder: QueryBuilder<UserDetailsModel, UserDetailsModel[]>,
+    payload: UserDetailsSearchUsersRequestDto,
+): void => {
+    if (payload.yearsOfExperience) {
+        void builder.where((subquery) => {
+            searchByYearsOfExperience(subquery, payload.yearsOfExperience);
+        });
+    }
+};
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/helpers/apply-filters/helpers/apply-experience-filter.ts]: The same provide only data which you will use

File: backend/src/bundles/user-details/helpers/apply-filters/helpers/apply-english-level-filter.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    if (payload.englishLevel) {
+        void builder.where((subquery) => {
+            searchByColumnValues(
+                subquery,
+                'englishLevel',
+                payload.englishLevel,
+            );
+        });
+    }
+};
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/helpers/apply-filters/helpers/apply-english-level-filter.ts]: The same provide only data which you will use

File: backend/src/bundles/user-details/helpers/apply-filters/helpers/apply-employment-type-filter.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
+import { type QueryBuilder } from 'objection';
+
+import { type UserDetailsSearchUsersRequestDto } from '../../../types/types.js';
+import { type UserDetailsModel } from '../../../user-details.model.js';
+
+const applyEmploymentTypeFilter = (
+    builder: QueryBuilder<UserDetailsModel, UserDetailsModel[]>,
+    payload: UserDetailsSearchUsersRequestDto,
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/helpers/apply-filters/helpers/apply-employment-type-filter.ts]: The same provide only data which you will use

File: backend/src/bundles/user-details/helpers/apply-filters/apply-filters.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    applyEnglishLevelFilter,
+    applyExperienceFilter,
+    applyHardSkillsFilter,
+    applyJobTitleFilter,
+    applyLocationFilter,
+    applyProfileNameFilter,
+    applyTypeSearchFilter,
+} from './helpers/filters.js';
+
+const applyFilters = (
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/helpers/apply-filters/apply-filters.ts]:
```suggestion
const applyAllFilters = (
```

**Review by MaksymGavatuk — CHANGES_REQUESTED** [review]

File: frontend/src/bundles/talent-onboarding/types/candidate-profile/candidate-profile.ts [hunk 11]
```diff
@@ -10,6 +10,7 @@ type FirstSectionDetails = {
     englishLevel: string;
     badges: typeof mockBadges;
     talentHardSkills?: string[];
+    hardSkills?: { id: string; name: string };
```
Comment (MaksymGavatuk) [inline frontend/src/bundles/talent-onboarding/types/candidate-profile/candidate-profile.ts]: Type please

File: frontend/src/bundles/search-candidates/types/seacrh-candidate-dto/seacrh-candidate-dto.ts [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    type ValueOf,
+} from 'shared/build/index.js';
+
+import { type DataStatus } from '~/bundles/common/enums/enums.js';
+
+type SeacrhCandidateDto = UserDetailsUpdateRequestDto & {
+    hardSkills?: {
+        id: string;
+        name: string;
+    }[];
```
Comment (MaksymGavatuk) [inline frontend/src/bundles/search-candidates/types/seacrh-candidate-dto/seacrh-candidate-dto.ts]: Type

File: frontend/src/bundles/search-candidates/store/actions.ts [hunk 13]
```diff
@@ -34,12 +37,9 @@ const getCandidateDetails = createAsyncThunk<
...
         try {
-            const userDetails =
-                await talentOnBoardingApi.getUserDetailsByUserId({
-                    userId: findPayload.userId,
-                });
-
-            return userDetails ?? null;
+            return (await talentOnBoardingApi.getUserDetailsByUserId({
+                userId: findPayload.userId,
+            })) as SeacrhCandidateDto;
```
Comment (MaksymGavatuk) [inline frontend/src/bundles/search-candidates/store/actions.ts]: Why you need to use casting ?

File: frontend/src/bundles/candidate-details/components/candidate-profile/candidate-profile.tsx [hunk 14]
```diff
@@ -95,7 +94,11 @@ const CandidateProfile: React.FC<Properties> = ({
         badges: mockBadges,
         preferredLanguages: data.preferredLanguages as string[],
         description: data.description as string,
-        talentHardSkills: hardskillsLabels,
+        talentHardSkills: isProfileCard
+            ? (data.hardSkills as { id: string; name: string }[]).map(
+                  (item) => item.name,
+              )
+            : hardskillsLabels,
```
Comment (MaksymGavatuk) [inline frontend/src/bundles/candidate-details/components/candidate-profile/candidate-profile.tsx]: Why use casting ?

File: backend/src/bundles/user-details/user-details.repository.ts [hunk 15]
```diff
@@ -218,12 +146,18 @@ class UserDetailsRepository implements Repository {
         );
 
         const searchResults = await query
-            .withGraphJoined('user')
-            .where('user.role', '=', UserRole.TALENT);
+            .withGraphJoined('[user, talentBadges.[badge]]')
+            .withGraphJoined('[user, talentHardSkills]')
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/user-details.repository.ts]: Why do you use separate withGraphJoined, if you need both of them use just .withGraphJoined('[user, talentHardSkills, talentBadges.[badge]]');

File: backend/src/bundles/user-details/user-details.model.ts [hunk 16]
```diff
@@ -91,6 +95,8 @@ class UserDetailsModel extends AbstractModel {
 
     public 'companyLogo'?: FileModel;
 
+    public 'badges'?: TalentBadgeEntity;
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/user-details.model.ts]: When you use graph fetching this type should be TalentBadgeModel

File: backend/src/bundles/user-details/user-details.entity.ts [hunk 17]
```diff
@@ -44,8 +49,14 @@ class UserDetailsEntity implements Entity {
 
     private 'preferredLanguages': ValueOf<typeof PreferredLanguage>[] | null;
 
+    public 'searchType': ValueOf<typeof SearchType>;
+
     private 'projectLinks': string[] | null;
 
+    private 'hardSkills'?: HardSkillsEntity[];
+
+    private 'badges'?: UserDetailsResponseBadgeDto[] | null;
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/user-details.entity.ts]: We don't need this field in entity

File: backend/src/bundles/user-details/user-details.entity.ts [hunk 18]
```diff
@@ -44,8 +49,14 @@ class UserDetailsEntity implements Entity {
 
     private 'preferredLanguages': ValueOf<typeof PreferredLanguage>[] | null;
 
+    public 'searchType': ValueOf<typeof SearchType>;
+
     private 'projectLinks': string[] | null;
 
+    private 'hardSkills'?: HardSkillsEntity[];
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/user-details.entity.ts]: We don't need this field in entity

File: backend/src/bundles/user-details/helpers/map-search-users-response-badges.ts [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+                level: badge.level,
+                badgeId: badge.badgeId,
+                isShown: badge.isShown,
+                userDetailsId: badge.userDetailsId,
+                badge: bsaBadge,
+            }).toObject();
+        }
+    });
+
+    const formattedBadges: (UserDetailsResponseBadgeDto | undefined)[] =
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/helpers/map-search-users-response-badges.ts]: Why we use undefined ?

File: backend/src/bundles/talent-badges/talent-badge.model.ts [hunk 20]
```diff
@@ -22,6 +23,8 @@ class TalentBadgeModel extends AbstractModel {
 
     public 'userDetailsId': string | null;
 
+    public 'badge'?: BSABadgeEntity;
```
Comment (MaksymGavatuk) [inline backend/src/bundles/talent-badges/talent-badge.model.ts]: model

File: backend/src/bundles/talent-badges/talent-badge.entity.ts [hunk 21]
```diff
@@ -15,6 +17,8 @@ class TalentBadgeEntity implements Entity {
 
     private 'userDetailsId': string | null;
 
+    private 'badge'?: BSABadgeEntity | null;
```
Comment (MaksymGavatuk) [inline backend/src/bundles/talent-badges/talent-badge.entity.ts]: We don't need this in enitity

File: backend/src/bundles/bsa-badges/bsa-badges.entity.ts [hunk 22]
```diff
@@ -1,13 +1,13 @@
...
+    public 'id': string | null;
 
-    private 'type': string;
+    public 'type': string;
 
-    private 'name': string;
+    public 'name': string;
 
-    private 'maxScore': number;
+    public 'maxScore': number;
```
Comment (MaksymGavatuk) [inline backend/src/bundles/bsa-badges/bsa-badges.entity.ts]: Why we change it to public ?

**Review by MaksymGavatuk — CHANGES_REQUESTED** [review]

**Comment by savinaDaria** [comment]
merged with #763 to apply&test proper candidate page handling

**Review by nikita-remeslov — APPROVED** [review]

**Pull request merged**